### PR TITLE
Builder for BranchPoints

### PR DESCRIPTION
**Pull request by agalbachicar**
# 🎉 New feature

Part of #10

## Summary

Implements the BranchPoint builder.

## Test it

Via unit tests (not yet).

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if it affects the public API)

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

**Note to maintainers**: this PR is in draft state. There are missing tests and it doesn't compile yet but it just aims to show how it would be used for preliminary discussions.

### CODE REVIEW

**Comment by agalbachicar** [comment]
@francocipollone I would appreciate your review on [90ccc36](https://github.com/maliput/maliput_sparse/pull/24/commits/90ccc36ee12c38e8d5249bda4cfa212fde99643e). I am mostly interested in knowing your thoughts on its use.

**Review by francocipollone — CHANGES_REQUESTED** [review]
lgtm overall.
I agree with the proposed api for the branchpoint building. I left some minor comments.

Populating the branchpoints is something that in general has to be implemented in each backend. I think that having this helper as agnostic to the backend as it is at maliput::base could be beneficial to avoid reimplementing code? Wdyt?

**Comment by agalbachicar** [comment]
This PR now depends on https://github.com/maliput/maliput/pull/524

File: include/maliput_sparse/builder/builder.h [hunk 1]
```diff
@@ -337,6 +376,12 @@ class RoadGeometryBuilder final {
   void SetJunction(maliput::common::Passkey<JunctionBuilder>,
                    std::unique_ptr<maliput::geometry_base::Junction> junction);
 
+  void SetBranchPoints(maliput::common::Passkey<BranchPointBuilder>,
```
[the remark this answers is not in the document]
Comment (agalbachicar) [reply include/maliput_sparse/builder/builder.h:420]: Done.

File: src/builder/builder.cc [hunk 2]
```diff
@@ -130,6 +132,144 @@ void JunctionBuilder::SetSegment(maliput::common::Passkey<SegmentBuilder>,
...
+}
+
+namespace {
+// Holds the A and B side sets of LaneEnds for a future maliput::geometry_base::BranchPoint.
+struct BranchPointSets {
+  std::set<BranchPointBuilder::LaneEnd> a_side{};
+  std::set<BranchPointBuilder::LaneEnd> b_side{};
+};
+
+// Proivdes maliput::api::BranchPointId with an increasing counter to make each ID
```
[the remark this answers is not in the document]
Comment (agalbachicar) [reply src/builder/builder.cc]: Done.

File: src/builder/builder.cc [hunk 3]
```diff
@@ -130,6 +132,144 @@ void JunctionBuilder::SetSegment(maliput::common::Passkey<SegmentBuilder>,
...
+class BranchPointIdBuilder {
+ public:
+  BranchPointIdBuilder() = default;
+
+  maliput::api::BranchPointId operator()() {
+    return maliput::api::BranchPointId(std::to_string(branch_point_count_++));
+  }
+
+ private:
+  int branch_point_count_{0};
```
[the remark this answers is not in the document]
Comment (agalbachicar) [reply src/builder/builder.cc]: Yes but it actually won't go negative until we go over int range. Do you think this is a problem?

File: src/builder/builder.cc [hunk 4]
```diff
@@ -176,5 +322,30 @@ void RoadGeometryBuilder::SetJunction(maliput::common::Passkey<JunctionBuilder>,
   junctions_.push_back(std::move(junction));
 }
 
+
+void RoadGeometryBuilder::SetBranchPoints(
+    maliput::common::Passkey<BranchPointBuilder>,
+    std::vector<std::unique_ptr<maliput::geometry_base::BranchPoint>>&& branch_points) {
```
[the remark this answers is not in the document]
Comment (agalbachicar) [reply src/builder/builder.cc:412]: Done.

File: src/builder/builder.cc [hunk 5]
```diff
@@ -176,5 +322,30 @@ void RoadGeometryBuilder::SetJunction(maliput::common::Passkey<JunctionBuilder>,
...
+
+std::unordered_map<maliput::api::LaneId, maliput::geometry_base::Lane*> RoadGeometryBuilder::GetLanes(
+    maliput::common::Passkey<BranchPointBuilder>) const {
+  std::unordered_map<maliput::api::LaneId, maliput::geometry_base::Lane*> lanes;
+  std::for_each(junctions_.begin(), junctions_.end(),
+      [lanes](const auto junction) {
+        for (int i = 0; i < junction->num_segments(); ++i) {
+          const maliput::api::Segment* segment = junction->segment(j);
+          for (int j = 0; j < segment->num_lanes(); ++j) {
+            maliput::geometry_base::Lane* lane = segment->lane(j);
```
[the remark this answers is not in the document]
Comment (agalbachicar) [reply src/builder/builder.cc]: Done.

File: src/builder/builder.cc [hunk 6]
```diff
@@ -130,6 +132,144 @@ void JunctionBuilder::SetSegment(maliput::common::Passkey<SegmentBuilder>,
   segments_.push_back(std::move(segment));
 }
 
+
+BranchPointBuilder& BranchPointBuilder::Connect(
+    const maliput::api::LaneId& lane_id_a, const maliput::api::LaneEnd::Which which_a,
+    const maliput::api::LaneId& lane_id_b, const maliput::api::LaneEnd::Which which_b) {
+  const LaneEnd lane_end_a = std::make_pair(lane_id_a, which_a);
+  const LaneEnd lane_end_b = std::make_pair(lane_id_a, which_b);
```
[the remark this answers is not in the document]
Comment (agalbachicar) [reply src/builder/builder.cc]: Changed the implementation.

File: src/builder/builder.cc [hunk 7]
```diff
@@ -130,6 +132,144 @@ void JunctionBuilder::SetSegment(maliput::common::Passkey<SegmentBuilder>,
...
+    BranchPointSets branch_point_set;
+    branch_point_set.a_side.emplace(lane_end);
+    branch_point_set.b_side = GetLaneEndSet(lane_end, lane_ends);
+    branch_point_sets->emplace(branch_point_set);
+  } else if (it_a_side != branch_point_sets->end()) {
+    auto b_side_set = GetLaneEndSet(lane_end, lane_ends);
+    it_a_side->b_side.insert(b_side_set.begin(), b_side_set.end());
+  } else if (it_b_side != branch_point_sets->end()) {
+    auto a_side_set = GetLaneEndSet(lane_end, lane_ends);
+    it_b_side->b_side.insert(a_side_set.begin(), a_side_set.end());
```
[the remark this answers is not in the document]
Comment (agalbachicar) [reply src/builder/builder.cc]: Changed the implementation.

**Review by francocipollone — APPROVED** [review]
LGTM! Nice job! This is a great addition!

File: test/builder/builder_test.cc [hunk 8]
```diff
@@ -320,6 +383,214 @@ GTEST_TEST(RoadGeometryBuilderTest, CompleteCase) {
...
+      maliput::api::LaneEnd(lane_b, start), maliput::api::LaneEnd(lane_c, start), maliput::api::LaneEnd(lane_d, start)};
+  const maliput::api::LaneEnd lane_end_a_finish(lane_a, finish);
+  // Evaluates the BranchPoints from the Lanes.
+  ASSERT_EQ(1, lane_a->GetConfluentBranches(start)->size());
+  ASSERT_EQ(0, lane_a->GetOngoingBranches(start)->size());
+  ASSERT_EQ(1, lane_a->GetConfluentBranches(finish)->size());
+  ASSERT_EQ(3, lane_a->GetOngoingBranches(finish)->size());
+  ASSERT_TRUE(lane_end_is_in(lane_a->GetOngoingBranches(finish)->get(0), expected_lane_ends));
+  ASSERT_TRUE(lane_end_is_in(lane_a->GetOngoingBranches(finish)->get(1), expected_lane_ends));
+  ASSERT_TRUE(lane_end_is_in(lane_a->GetOngoingBranches(finish)->get(2), expected_lane_ends));
```
[the remark this answers is not in the document]
Comment (agalbachicar) [reply test/builder/builder_test.cc:566]: You are right. I did not think of it. Let me address that in a follow up PR.

File: src/builder/builder.cc [hunk 9]
```diff
@@ -130,6 +132,229 @@ void JunctionBuilder::SetSegment(maliput::common::Passkey<SegmentBuilder>,
...
 
+BranchPointBuilder& BranchPointBuilder::Connect(const maliput::api::LaneId& lane_id_a,
+                                                const maliput::api::LaneEnd::Which which_a,
+                                                const maliput::api::LaneId& lane_id_b,
+                                                const maliput::api::LaneEnd::Which which_b) {
+  const LaneEnd lane_end_a{lane_id_a, which_a};
+  const LaneEnd lane_end_b{lane_id_b, which_b};
+  //@{ Asserts that we are not inserting duplicate keys.
+  auto range_a_b = lane_ends_.equal_range(lane_end_a);
+  auto range_b_a = lane_ends_.equal_range(lane_end_b);
```
[the remark this answers is not in the document]
Comment (agalbachicar) [reply src/builder/builder.cc]: Yes, good catch!

**Review by francocipollone — APPROVED** [review]
:1st_place_medal:

**Pull request merged**